server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -62,59 +66,88 @@ public void addServerAndUpdateSegment(
   {
     synchronized (this) {
       this.segment.set(segment);
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.computeIfAbsent(priority, p -> new HashSet<>());
+      Set<QueryableDruidServer> priorityServers = new HashSet<>();
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        priorityServers = realtimeServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      } else if (server.getServer().getType() == ServerType.HISTORICAL) {
+        priorityServers = historicalServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      }
       priorityServers.add(server);
     }
   }
 
   public boolean removeServer(QueryableDruidServer server)
   {
     synchronized (this) {
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.get(priority);
-      if (priorityServers == null) {
-        return false;
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        return realtimeServers.computeIfPresent(
```

Comment:
Suggested to choose "servers" map based on the server type, and then implement the removal logic only once